### PR TITLE
TEM Store factorizations in RAM

### PR DESCRIPTION
**Pull request by dccowan**
In this PR, we add the option to store factorizations in RAM. The ability to read/write factorizations can be handled as part of another project. Just like the FEM problem, this is controlled using the 'forward_only' kwarg

### CODE REVIEW

**Comment by dccowan** [comment]
Is there a particular version of black that must be used? I already ran 'black simulation.py' but it's failing the black test on Azure. Please advise.

**Comment by jcapriot** [comment]
SimPEG currently runs black at v21.7b0

**Comment by jcapriot** [comment]
There’s a lot of logic here to determine the step length indices because you’re storing factored Ainv’s in an array. Store them in a dictionary that is indexed by `dt` and you can simplify this significantly.

**Comment by dccowan** [comment]
> There’s a lot of logic here to determine the step length indices because you’re storing factored Ainv’s in an array. Store them in a dictionary that is indexed by `dt` and you can simplify this significantly.

Oh wow, didn't know you could use a float as a dictionary key. Ya, that's more compact.

File: SimPEG/electromagnetics/time_domain/simulation.py [hunk 1]
```diff
@@ -86,36 +100,79 @@ def fields(self, m):
...
+        f[:, self._fieldType + "Solution", 0] = self.getInitialFields()  # mesh x n_src
+
+        # Store Factorizations
+        if self.forward_only is False:
+
+            # Clean factorizations for preexisting model.
+            if hasattr(self, "Ainv"):
+                {k: v.clean() for k, v in self.Ainv.items()}
+            else:
+                self.time_steps = self.dt_threshold * np.round(
```
Comment (jcapriot) [inline SimPEG/electromagnetics/time_domain/simulation.py:119]: This will fail very quickly if dt_threshold is 0

Comment (dccowan) [reply SimPEG/electromagnetics/time_domain/simulation.py:119]: K. So let's update the setter so that dt_threshold must be larger than 0. That work?

**Pull request closed without merging**